docs/AllocatorClaims.md

```diff
@@ -0,0 +1,120 @@
+
+Note: This approach can be orthogonal to others and so we should implement it anyway.
+
+Possible approach 1: Hazard pointers
+------------------------------------
+
+Hazard pointers provide inspiration for the first possible solution.
+This approach involves having a per-compartment list of held objects that the allocator must consult before freeing an object.
+The allocator would provide an API that allowed a compartment to register a table containing a (bounded) array of held objects and an allocation capability.
+On deallocation, the allocator would need to traverse all such lists to determine whether an object is on any of them.
+If the object is found on a list then the metadata should be updated to mark it as owned by the allocation capability associated with the list.
+
+This is attractive for the fast-path case, because the compartment wishing to claim an object for the duration of a call needs only to insert a pointer to it into a pre-registered array.
+It is unclear whether it is possible to use this interface securely.
+In particular, a malicious compartment could allocate a large object and pass a pointer to a small sub-object to the callee.
+The callee has no mechanism (other than calling the allocator) to know whether it holds a pointer to an entire object.
+
+Possible approach 2: Explicit claims
+------------------------------------
+
+The approach in CheriOS is similar to reference counting.
+Compartments may explicitly claim memory and objects may not be deallocated until all claims have been dropped.
+Importantly, the CheriOS model handles resource constraints: when you claim an object, it counts towards your quota.
+
+The down side of this model is that it requires tracking a lot of state.
+Straight reference counting is not sufficient because references can be owned by mutually distrusting entities.
+With a simple reference counting model, the original attack is still possible:
+
+1. A allocates an object.
+2. A passes a pointer to the object to B.
+3. B increments the reference count.
+4. A decrements the reference count twice from another thread.
+5. B traps.
+
+Each reference to the object is now state that must be tracked and, most likely, will require an allocation.
+We currently have 14 bits spare in the allocation header (slightly more if we shrink the allocator ID, which does not actually require a full 16 bits, since a 16-bit allocator ID would require 2 MiB of RAM to hold all of the corresponding allocation capabilities, leaving no space for the heap) and so could store a heap-start-relative (shifted for alignment) pointer to the head of a linked list of claims.
+In the common case, this field will be 0 (the start of the heap cannot be allocated) and so free can skip it.
+
+The proposed solution involves constructing a linked list of structures holding an allocator ID, a saturating reference count, and a next pointer.
```

Comment:
What happens on saturation, the object becomes immortal?

Reply:
Yup, if you claim an object 2^32 times (or possibly 2^16 times), it's yours forever and you get to keep it.

For reference, Smalltalk-80 did the same thing, only with 7, not 2^16, and that worked surprisingly well.